### ISSUE TITLE
Forgot password policy

### ISSUE THREAD

**Issue by DibranMulder** [Azure-Samples/active-directory-b2c-javascript-msal-singlepageapp#9]
When using a combined signup - signin policy in Azure B2C, users have to handle the forgot password scenario themselves. Please take a look at the [documentation](https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-reference-policies).

> How do I link a sign-up or sign-in policy with a password reset policy?
> When you create a sign-up or sign-in policy (with local accounts), you see a Forgot password? link on the first page of the experience. Clicking this link doesn't automatically trigger a password reset policy.
> Instead, the error code AADB2C90118 is returned to your app. Your app needs to handle this error code by invoking a specific password reset policy. For more information, see a sample that demonstrates the approach of linking policies.

I managed to redirect the page to the forgot password policy when that error with error_description AADB2C90118 occurs. However when the user then successfully resets his or hers password I'm not able to login the user, it seems that the authority's do not match since the policy names are included in the authority names. Can you please include the forgot password scenario in your samples?

**Comment by parakhj** [comment]
We are working to update our samples, but taking a quick stab on one way you can do this is:
in applicationConfig variable, add `authorityPR` property (PR = Password Reset) with the same value as authority, but replacing the policy at the end of the line with a password reset policy. Then adding the following line:

`
var clientApplicationPR = new Msal.UserAgentApplication(applicationConfig.clientID, applicationConfig.authorityPR, function (errorDesc, token, error, tokenType) {
            // Called after loginRedirect or acquireTokenPopup
        });
`
Wherever you are doing the error handling for the error code, call `clientApplicationPR.acquireTokenPopup(...)`

**Comment by DibranMulder** [comment]
How is this supposed to work? I want both login as password reset functionality available. Should I create two UserAgentApplications with both their own authority? This is my workflow:

1. User heads to Azure B2C page with redirectLogin.
```
    /**
     * Redirects the user to the login page
     */
    public login() {
        this.clientApplication.loginRedirect(this.applicationConfig.b2cScopes, `ui_locales=${this.$translate.use()}`);
    }
```
2. User clicks forgot password button
3. Azure B2C redirects back to my application with an error and error_description containing AADB2C90118.
```
    /**
     * Checks if the storage contains an error and and the errors contains the forgot password code.
     * @returns return true if the user is not redirected, false if the user forgot its password and the browser has to redirect.
     */
    public redirectIfForgotPassword() {
        let error = this.storage.getItem("error");
        // An error should be there
        if (error) {
            // And the description should contain: AADB2C90118
            let errorDesc = this.storage.getItem("error_description");
            if (errorDesc.indexOf("AADB2C90118") > -1) {
                // Set the authority to the forgot password
                this.clientApplication.authority = this.applicationConfig.forgotPasswordAuthority;
                // and head there.
                this.login();
                return false;
            }
        }
        return true;
    }
```
4. I redirect to the forgot password page by setting the authority to the Forgot Password policy
`this.clientApplication.authority = this.applicationConfig.forgotPasswordAuthority;`
5. User changes its password and Azure B2C redirects back to the application with a valid token
6. My application gets loaded but when I perform an acquireTokenSilent I get the following error:

> Refused to display 'https://login.microsoftonline.com/te/{tenant}.onmicrosoft.com/b2c_1_signup-signin/oauth2/v2.0/authorize?response_type=token&scope=https%3A%2F%2F{tenant}.onmicrosoft.com%2F{appname}%2Fread%20openid%20profile&client_id=968a7cc1-da99-40fd-9cb6-7a01f5545e48&redirect_uri=https%3A%2F%2Flocalhost%3A44399&state=eef3516e-f8c4-4e10-8deb-02a4473a1f0c&nonce=e31f3653-8bb0-471e-82e3-24c96beefb9f&client_info=1&x-client-SKU=MSAL.JS&x-client-Ver=0.1.1&client-request-id=10db8ff1-c2f6-45f2-ad15-dd69597937bf&prompt=none&domain_req=627de5e7-f8a1-4584-8095-a6acd06a9ce1&login_req=45807b24-89c9-4414-a303-14105398ad65-b2c_1_ccs_pwd_reset&domain_hint=organizations' in a frame because it set 'X-Frame-Options' to 'deny'.

> GET https://login.microsoftonline.com/te/{tenant}.onmicrosoft.com/b2c_1_signup-signin/oauth2/v2.0/authorize?response_type=token&scope=https%3A%2F%2F{tenant}.onmicrosoft.com%2F{appname}%2Fread%20openid%20profile&client_id=968a7cc1-da99-40fd-9cb6-7a01f5545e48&redirect_uri=https%3A%2F%2Flocalhost%3A44399&state=eef3516e-f8c4-4e10-8deb-02a4473a1f0c&nonce=e31f3653-8bb0-471e-82e3-24c96beefb9f&client_info=1&x-client-SKU=MSAL.JS&x-client-Ver=0.1.1&client-request-id=10db8ff1-c2f6-45f2-ad15-dd69597937bf&prompt=none&domain_req=627de5e7-f8a1-4584-8095-a6acd06a9ce1&login_req=45807b24-89c9-4414-a303-14105398ad65-b2c_1_ccs_pwd_reset&domain_hint=organizations net::ERR_BLOCKED_BY_RESPONSE

Looks like an token is requested at the b2c_1_signup-signin policy but the acquired user information is retrieved from the other policy. I'm kinda stuck now.

**Comment by rahulbisht225** [comment]
@DibranMulder  Have you got any solution for this issue. I am also facing the same issue.I  thought of logging out once i catch this error and then login again.

Please help me i am stuck at this.

Thanks

**Comment by DibranMulder** [comment]
@rahulbisht225 I'm also stuck at this... Seems like its not implemented in the msal library, haven't found any other lib that can take care of this either. At msal they closed my issue stating its not something they are going to implement right now.
https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/111

The workaround I currently take is logging out when the acquireTokenSilent fails. I don't use popups since every modern browser just blocks them so I think its save to logout when the acquireTokenSilent fails. It also fails when the user is logged in via the forget password policy, so it covers also that scenario.

**Comment by jmprieur** [comment]
@DibranMulder : we wrote we are going to consider it (not yet implement)
BTW, don't you want to do acquireTokenRedirect when acquireTokenSilent fails?

**Comment by DibranMulder** [comment]
@jmprieur Thanks for the hint, I'll check the behavior.

**Comment by mikelunn** [comment]
@parakhj Is there a timeline of when these samples may be updated. I really would like to use Azure B2C for a new large scale enterprise application. I have the forgot password functionality working. However, the user experience is not ideal.

**Comment by Chakravarti-Baratam** [comment]
@DibranMulder , did you get it working? I am also facing the same issue.

**Comment by DibranMulder** [comment]
@Chakravarti-Baratam Not yet.. I still take the workaround below, when the next release of msal is there ill try to acquireTokenRedirect when acquireTokenSilent fails. Hopefully I can find some workflow to login the user when he changes his password.

> The workaround I currently take is logging out when the acquireTokenSilent fails. I don't use popups since every modern browser just blocks them so I think its save to logout when the acquireTokenSilent fails. It also fails when the user is logged in via the forget password policy, so it covers also that scenario.

**Comment by Chakravarti-Baratam** [comment]
@DibranMulder , could you share your piece of code? Where are you calling this function redirectIfForgotPassword'? Is it in your redirect callback function?

My code looks like below:

window.onload = function () {
  window.clientApplication = new Msal.UserAgentApplication(MSAL_CLIENTID, AUTHORITY,
  authCallback, {cacheLocation: 'localStorage', redirectUri: MSAL_REDIRECTURI});
  var user = window.clientApplication.getUser();
  if (user) {
  
  }else{
    if (!window.clientApplication.isCallback(locationHash) && window.parent === window && !window.opener) {
      window.clientApplication.authority = AUTHORITY;
      window.clientApplication.redirectUri = MSAL_REDIRECTURI;
      window.clientApplication.loginRedirect([MSAL_SCOPE]);
    }else if(locationHash.indexOf("AADB2C90118") > -1){
      window.clientApplication.authority = AUTHORITY_PR;
      window.clientApplication.redirectUri = MSAL_REDIRECTURI;
      window.clientApplication.loginRedirect([MSAL_SCOPE]);
    }    
  }  
}

My callback is not doing much.

function authCallback(errorDesc, token, error, tokenType) {
  if(token) {
  }
  else if (errorDesc || error) {
    console.error(error + ':' + errorDesc);    
  }  
}

**Comment by mikelunn** [comment]
@Chakravarti-Baratam You'll have to handle the forgot password redirect in the authCallback function and check if the errorDesciption string contains the error code. 
`
if (errorDesc.indexOf('AADB2C90118') > -1)
{redirect();}`

**Comment by Chakravarti-Baratam** [comment]
@mikelunn , I am getting null value for UserAgentApplication object in authCallback function. So unable to redirect from authCallback function. Could you share your snippet of code, please.

**Comment by jcarloslr10** [comment]
@Chakravarti-Baratam I got it to work with this code for the UserAgentApplication object

```
this._app = new Msal.UserAgentApplication(this._tenantConfig.clientId, this._tenantConfig.authority,
            (errorDesc, token, error, tokenType) => {
                this._storageService.save(this._b2cErrorKey, errorDesc);
            });
```
and

```
login(): any {
        return this._app.acquireTokenSilent(this._tenantConfig.scopes)
            .then(accessToken => {
                this._isLogged = true;
                this._access_token = accessToken;
                return accessToken;
            }, error => {
                this.redirectIfForgotPassword();
                return this._app.loginRedirect(this._tenantConfig.scopes, `ui_locales=${this._languageService.currentLanguage}`)
                    .then(accessToken => {
                        this._access_token = accessToken;
                        return accessToken;
                    }, err => {
                        console.error(err);
                    });
            });
    }

redirectIfForgotPassword(): void {
        const errorDesc = this._storageService.get(this._b2cErrorKey);
        if (errorDesc && errorDesc.indexOf('AADB2C90118') > -1) {
            const authUrlSplit = this._tenantConfig.authority.split('/');
            const authorityUrl =
                `${authUrlSplit.slice(0, authUrlSplit.length - 1).join('/')}/${environment.b2c.passwordResetPolicy}`;

            this._app.authority = authorityUrl;
            this._storageService.remove(this._b2cErrorKey);
            
            this.login();
        }
    }
```
for login and redirect methods. This will allow you to change the policy without getting null value for UserAgentApplication object. I'm not found any better solution. @parakhj when will you have the right examples?

Thanks!!

**Comment by Chakravarti-Baratam** [comment]
@parakhj , could you please update the sample for this scenario?

**Comment by spottedmahn** [comment]
>`if (errorDesc.indexOf('AADB2C90118') > -1) {redirect();}`

Hello @mikelunn - what does your `redirect()` function look like?

**Comment by spottedmahn** [comment]
Hello @jcarloslr10 - thanks for the code!  Feels very hacky... I sure hope they respond/address this.

It seems like msal.js has not taken into consideration the concept of policies in B2C and that it assumes only one policy.

**Comment by mikelunn** [comment]
@spottedmahn Here it is. Changed name to onTokenCallback. But you essentially need to re-instantiate the UserAgentApplication object and pass in your password reset policy.

```
private _onTokenCallback = (errorDesc: any, token: any, error: any, tokenType: any) => {
        if (token) {
            this.accessToken = token;
          }
          if (errorDesc) {
            if (errorDesc.indexOf('AADB2C90118') > -1) {
              //  forgot
              console.log(errorDesc);
              this.userAgentApplication = new UserAgentApplication(
                AppConfig.tenantConfig.clientID,
                AppConfig.tenantConfig.passwordReset,
                this._onTokenCallback
              );
              this.signIn();
            }
          }
     }
```

Here is my complete class which kind of wraps msal.js. Keep in mind this is for angular/typescript

```
import { Injectable } from '@angular/core';
import { AppConfig } from '../app.config';
import {UserAgentApplication} from 'msal';
@Injectable()
export class MsalService {
    accessToken: string;
    userAgentApplication: UserAgentApplication = null;

    constructor() {
        this.userAgentApplication = new UserAgentApplication(
            AppConfig.tenantConfig.clientID,
            AppConfig.tenantConfig.signInSignUp,
            this._onTokenCallback

        )
     }
     signIn(): void {
        const clientApp = this.userAgentApplication;
        const tenantConfig = AppConfig.tenantConfig;
        clientApp.loginRedirect(tenantConfig.b2cScopes);
     }
     signOut(): void {
        this.userAgentApplication.logout();
     }
     isAuthenticated(): boolean{
        return this.userAgentApplication.getUser() != null;
     }

     getToken(){}

     private _onTokenCallback = (errorDesc: any, token: any, error: any, tokenType: any) => {
        if (token) {
            this.accessToken = token;
          }
          if (errorDesc) {
            if (errorDesc.indexOf('AADB2C90118') > -1) {
              //  forgot
              console.log(errorDesc);
              this.userAgentApplication = new UserAgentApplication(
                AppConfig.tenantConfig.clientID,
                AppConfig.tenantConfig.passwordReset,
                this._onTokenCallback
              );
              this.signIn();
            }
          }
     }
}
```

**Comment by AnnaShk** [comment]
Hi,

I've used the mentioned by @mikelunn solution (commented on Nov 29, 2017) for user forget password.

As a result I have this request (from dev tools): 

> **Request URL:** https://login.microsoftonline.com/${configSettings.tenant}/oauth2/v2.0/authorize?p=${configSettings.passwordResetPolicy}&client_id=${configSettings.clientID}&nonce=defaultnonce&redirect_uri=http%3a%2f%2flocalhost%3a4200&scope=openid&response_type=id_token&prompt=login&/v2.0/.well-known/openid-configuration
> **Request Method:** GET
> **Status Code:** 200 OK
> **Referrer Policy:** no-referrer-when-downgrade
> 
> **Provisional headers are shown**
> **Origin:** http://localhost:4200
> **Referer:** http://localhost:4200/home
> **User-Agent:** Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.87 Safari/537.36


Issues:
1. Could you please help me to find why Request Headers are almost empty? So the request fails with no CORS headers (No 'Access-Control-Allow-Origin' header is present on the requested resource).

2. The msal adding /v2.0/.well-known/openid-configuration in the end which breaks the request URL. Is there any way to fix that? Currently I just separate it with & at the end of authority.


Thank you 

My AuthenticationService:
```
const passwordresetAuthority = "https://login.microsoftonline.com/${configSettings.tenant}/oauth2/v2.0/authorize?p=${configSettings.passwordResetPolicy}&client_id=${configSettings.clientID}&nonce=defaultNonce&redirect_uri=http%3A%2F%2Flocalhost%3A4200&scope=openid&response_type=id_token&prompt=login&";
export class AuthenticationService {

    private authority = "https://login.microsoftonline.com/tfp/${configSettings.tenant}/${configSettings.signUpSignInPolicy}";

    private clientApplication: Msal.UserAgentApplication;
    private token: Token;

    constructor() {
        this.clientApplication =
            new Msal.UserAgentApplication(
                configSettings.clientID,
                this.authority,
                authCallback,
                {
                    redirectUri: window.location.origin,
                }
            );
    }
    public login(): void {
        this.clientApplication.loginRedirect(configSettings.b2cScopes);
    }
    public logout(): void {
        this.clientApplication.logout();
    }
    public acquireTokenSilent(): Promise<string> {
        return this.clientApplication.acquireTokenSilent(configSettings.b2cScopes)
            .then(token => {
                return token;
            }).catch(error => {
                console.error("error getting access token: " + error);
                return Promise.resolve("");
            });
    }
}

function authCallback(errorDesc: any, token: any, error: any, tokenType: any): void {        
    if (errorDesc || error) {
        if (errorDesc.indexOf("AADB2C90118") > -1) {            
            let clientApp = window.msal as Msal.UserAgentApplication;            
            clientApp.authority = passwordresetAuthority;

            clientApp.loginRedirect(configSettings.b2cScopes);
        }
        else {
          console.log(error + ":" + errorDesc);
        }
    }
}

**Comment by chriseugenerodriguez** [comment]
@mikelunn What is in your appconfig i.e passwordReset variable?

**Comment by mikelunn** [comment]
@chriseugenerodriguez The passwordReset variable is  the name of the PasswordReset Policy for your tenant.  Below is an example.
https://login.microsoftonline.com/tfp/My-Tenant-Name/B2C_1_MyPasswordResetPolicy

**Comment by chriseugenerodriguez** [comment]
@mikelunn thanks, looks good.

**Comment by SimeonBanev** [comment]
@mikelunn Your example works for me but after the user resets their password and they get redirected back to the app if I call acquireTokenSilent I get the same error that @DibranMulder described:

 - 'Refused to display url in a frame because it set 'X-Frame-Options' to 'deny'. 

Is anyone else having this issue and did you manage to find a solution?

**Comment by ghost** [comment]
@SimeonBanev same here, so i use also this workaround:

`
this.broadcastService.subscribe("msal:acquireTokenFailure", (payload) => {
 console.log("logging out");
      this.logout();
    });
`

**Comment by danielekbom** [comment]
I am using the solution posted by mikelunn. It works sometimes. About 50% of the times when I click on "Forgot your password" link it redirects me back to the login page. Is someone else having this issue or know what the problem might be? Thanks!

**Comment by tank104** [comment]
Hi @mikelunn  I noticed that your msal.service.ts misses a lot of the code that is in the nuget package:
https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-angular/src/msal.service.ts

Is all that code not important?

**Comment by tank104** [comment]
I am also have trouble getting your example to work @mikelunn .  
It will always go to the signin page, even after it successfully gets to the code where we change authority to password reset.

Should I still have this in the NgModule:
```
    MsalModule.forRoot({
      clientID: environment.msalClientID,
      authority: environment.msalAuthority + environment.msalSigninPolicy,
    }),
```

It seems whatever is setup there overwrites whatever I pass in to the authorityUsed here:

`    this.userAgentApplication = new UserAgentApplication(environment.msalClientID, authorityUsed, this._onTokenCallback);`

Actually just to correct that - every so often it works, but no discernible pattern as to when that is.  It looks like it redirects twice looking at fiddler.  First to password reset, then immediately to login.

**Comment by mikelunn** [comment]
I moved my auth logic to the server side with their .net core libraries.
I'm not certain if my code example is still relevant. I would defer to the
Azure example repository that you posted.

**Comment by tank104** [comment]
Do you happen to have an example of yours above working you could share? I really would like to figure out why mine only works sometimes, and why it trys to go to both aignup and password reset pages at the same time.

**Comment by tank104** [comment]
Ok so I managed to get it working, its a bit complicated, and unfortunately meant changing the "@azure/msal-angular" npm package.
In brief what I did was copied the package source into my folder "app/msal-angular", then I made a pass-through class around MsalService.  This wrapper class created MsalService, and passed in a callback onTokenCallback.  The code within that was this;

```
  private _onTokenCallback = (originalMsalService: OriginalMsalService, errorDesc: any, token: any, error: any, tokenType: any) => {
    if (errorDesc) {
      console.log(errorDesc);
      if (errorDesc.indexOf("AADB2C90118") > -1) {
        originalMsalService.getCacheStorage().removeItem(Constants.loginRequest);
        originalMsalService.authority = environment.msalAuthority + environment.msalPasswordResetPolicy;
        originalMsalService.loginRedirect();
      }
    }
  };
```

MsalService was changed so that the callback had the signature:
```
export declare type tokenReceivedCallback = (
  originalMsalService: MsalService,
  errorDesc: string,
  token: string,
  error: string,
  tokenType: string,
  userState: string
) => void;

```
Then in the method processHash when I call the callback I pass in this (MsalService) so that it can change the properties above. 

Been a real challenge to get to this point, as its my first Typescript exercise, but it works, and is as neat as I possibly can make it.


If the library was changed to be able to pass back MsalService in the callback, life would be very simple.

**Comment by tank104** [comment]
Actually there is an easier/cleaner fix. 
Copy the msal-angular source code to your app.  Add a new option in msal-config called passwordResetAuthority, then change the code below to add in the first if statement in msal.service.ts:

```
          if (errorDescription && errorDescription.indexOf("AADB2C90118") > -1) {
            // This error is returned by ADB2C for forgotten password
            this.authority = this.passwordResetAuthority;
            this.loginRedirect();
          } else if (this._navigateToLoginRequestUrl) {
            let loginStartPage = this._cacheStorage.getItem(Constants.loginRequest);
            this._cacheStorage.setItem(Constants.urlHash, hash);
            if (typeof loginStartPage !== "undefined" && loginStartPage && loginStartPage.length !== 0) {
              // prevent the current location change and redirect the user back to the login start page
              this._logger.verbose("Redirecting to start page: " + loginStartPage);
              window.location.href = loginStartPage;
            }
```

**Comment by MaximBalaganskiy** [comment]
Another option is integrating password reset in the same signin-signup policy. Then the id token, issued after a reset will belong to the same authority and can be used to get access tokens. You will still need to catch "access_denied" errors and redirect to this super combined policy with a token (InputTokenFormat = JSON or JWT) which has to contain a "trigger" claim directing the user flow.

**Comment by jennyf19** [comment]
@MaximBalaganskiy Do you have documentation or code which shows how to do this? Seems like a quick fix and easiest way around this issue.

**Comment by MaximBalaganskiy** [comment]
@jennyf19 [This](https://github.com/Azure-Samples/active-directory-b2c-advanced-policies/blob/master/wingtipgamesb2c/Policies/b2ctechready.onmicrosoft.com_B2C_1A_invitation.xml) policy demonstrates how to pass additional claims to the user flow. I couldn't find any examples on how to use JSON though - must be similar and simpler - no need to sign. In the user journey you can activate different "paths" depending on a claim presence.

**Comment by DannyRyman** [comment]
I have done some investigation regarding the "Refused to display url in a frame because it set 'X-Frame-Options' to 'deny'" issue when logging into the application after a password reset. I think AAD has a couple of issues in how it has implemented this functionality. The first thing is that they are not storing a SSO cookie as part of the password reset policy. This is exasperated by a second issue that the prompt=none querystring of the authorize endpoint is being ignored, and instead shows the login screen.

So what is happening is that you go through the Password Reset flow, which returns a success callback, allowing the user to log into the application. However, unlike the login flow, it has not set a SSO cookie. When the application attempts to get an access token, it uses an IFrame (to get around lack of CORS headers) to call the authorize endpoint with the prompt=none. Because there in no SSO cookie, it should return an error; unfortunately I does not and instead returns a 200 response with the login page. It then tries to load the login form into the IFrame, but due to the X-Frame-Options - deny header, it is showing the aforementioned error.

If this analysis is correct, then unless MS fix the issue, I think the workaround is to force a password login after a reset (as mentioned in the comments above).

**Comment by SujitGodavarti** [comment]
I use MSAL for authentication in my angular app and Sign in and Sign up work perfectly. I am trying to implement forgot password functionality now. I created the necessary policy in the azure portal. When I click on 'Forgot your password' link, I encounter a CORS error mentioned below. 

Access to XMLHttpRequest at 'https://myb2ctest.b2clogin.com/myb2ctest.onmicrosoft.com/oauth2/v2.0/authorize?p=my_b2c_passwordreset_policy&client_id=****************&nonce=defaultnonce&redirect_uri=http%3a%2f%2flocalhost%3a4200%2fpages%2fforgetpassword&scope=openid&response_type=id_token&prompt=login&/v2.0/.well-known/openid-configuration' from origin 'http://localhost:4200' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

I couldn't figure out what exactly is causing this issue. Could you please help me this issue?

**Comment by spottedmahn** [comment]
Hi @SujitGodavarti - I would personally open another issue as this issue related but feels different to me.

**Comment by dluque92** [comment]
My problem is after forgot password/editProfile the callback don't return into the correct callback, because in this case the state is not included in the initialization, and also the callback is outside the initialization too.

Function onTokenCallback changed again, 

https://docs.microsoft.com/bs-latn-ba/azure/active-directory/develop/msal-js-initializing-client-applications

Now is like this: 
````
// create UserAgentApplication instance
const myMSALObj = new UserAgentApplication(config);

myMSALObj.handleRedirectCallback(authCallback);

function authCallback(error, response) {
    //handle redirect response
}

// And config is like this: 
config: Configuration = {
    auth: AuthOptions,
    cache?: CacheOptions,
    system?: SystemOptions,
    framework?: FrameworkOptions
};

**Comment by ErikAndreas** [comment]
Having the same problem as @dluque92, as well as with update user profile, similar to https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/440. See also my issue on b2c docs https://github.com/MicrosoftDocs/azure-docs/issues/41013

I've found this PR https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/915, could someone shed some light on

1. will this PR fix these issues?
2. in which release will this PR be included?
3. when will that release be released?

**Comment by jennyf19** [comment]
fyi: @DarylThayil @sameerag

**Comment by rakeshrajagopal24** [comment]
Folks,
We are getting same error in our envrionment (see below). "*error=access_denied&error_description=AADB2C90118%3a+The+user+has+forgotten+their+password.%0d%0aCorrelation+ID%3a+f5af0d19-7e04-4473-93d8-1e4f58304951%0d%0aTimestamp%3a+2019-12-18+07%3a45%3a30Z%0d%0alogin_hint%3a+rka%0d%0a*" 

We have Okta as passthrough and Azure B2C as IDP. Thus when we get this error we are being redirected back to application page with following error. "*access_denied&error_description=User+canceled+the+social+login+request*"

Now, we don't have any control to handle this error at Okta (as Okta is working as intermediate between Azure B2C and end application) and rather it shall be handled at Azure B2C end itself. Please suggest where/how shall we add this suggested chunk of code to handle this error.

Thanks..!!

**Comment by nlz242** [comment]
To add to @rakeshrajagopal24 's comment... 
We are a company that provide a web app to different orgs (cities, non-profits, etc), this web app supports OpenID Connect so that our clients (orgs) can configure it to work with their IdP of choice. Our implementation follows the OpenID Connect standards, we strive to keep it as standard as possible so that it works everywhere. Recently, we tryed to get our app to work with Azure B2C and found out that we couldn't simply delegate authentification to the IdP, we had deal with this "Forgot Passwords redirects to the caller" thing... Now we have managed the callback from Azure B2C with the specific Azure B2C error code so that we could then redirect the user to the other URLs found in the other discovery document. That all seems pretty janky and very specific to azure. 

I think a fully featured "Flow" should exists for B2C Clients who aren't in control of the site that will B2C as IdP (just like the case of rakeshrajagopal24 with otka). A Flow that has Sign up and sign in v2, Password reset v2 and Profile editing v2 all in one.

**Comment by jennyf19** [comment]
@nlz242 @rakeshrajagopal24 Thanks for the feedback. B2C is working on improving the developer experience here, and having the service handle the password reset directly from the dialogue is on the backlog. 

cc: @valnav @nickgmicrosoft @ktsakas to add more info here.

**Comment by Ethan-Arrowood** [comment]
Is anyone able to redirect within the popup?

**Comment by jasonnutter** [comment]
@Ethan-Arrowood Are you referring to this? https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/1198

**Comment by Ethan-Arrowood** [comment]
I believe so. In order to swap between authorities I needed to set the authority for the instance then call the login function again. Ideally I could just redirect the user from sign in/sign up to password_reset in one popup but 🤷‍♀️

**Comment by adomrockie** [comment]
this is a basic functional requirement for any user login process which should have been thought about before even releasing this. I do not mind using the workaround but since 2017 this ticket is still open, guys sort this out

**Comment by derisen** [comment]
@adomrockie apologies for this. Better support for multiple policies and other B2C features is on its way. In the meantime, you can check out the `derisen-refactoring` branch where a workaround is demonstrated (in `authPopup.js` file).

**Comment by jennyf19** [comment]
@derisen can your branch be merged into master at some point so the functionality is there (even as a work around)?

**Comment by derisen** [comment]
@jennyf19 will do that as soon as msal.js v1.3.0 is released (which fixes an issue with using the workaround with the redirect flow)

**Comment by jennyf19** [comment]
Thanks @derisen ...do you have an ETA on that?

**Comment by derisen** [comment]
@jennyf19 you're welcome : ) I'm hoping for by the end of this week, but the timeline for release is not clear yet.

**Comment by jasonnutter** [comment]
@jennyf19 @derisen I'm planning to have 1.3 available today.

**Comment by jasonnutter** [comment]
Update: 1.3.0 is available via npm and cdn.
